### ISSUE TITLE
Very noticeable input lag

### ISSUE THREAD

**Issue by Tasosgemah** [GLupeN64/GLupeN64#156]
It seems like, for me, Glupen has much more input lag compared to the regular Mupen core.

I tested the emulator with GPU sync - ON and GPU frames = 0, which should be the optimal settings for the lowest input lag. I loaded Mario 64 and i used the frame advance feature to count how many frames it takes for Mario to jump after i press the jump button.

With the above optimal settings, it took Mario 5 frames to leave the ground. This is pretty high IMO.

In comparison, in the regular Mupen core, Mario only needs 2 frames. 

This is a pretty big difference so maybe there is a problem with Glupen? Playing with it just feels less responsive. Otherwise it's pretty great and probably the best core so far. Can this be improved to shave off a couple of frames at least?

**Comment by loganmc10** [comment]
Can you test mupen64plus-libretro with the angrylion video plugin?

**Comment by loganmc10** [comment]
In my testing, using mupen64plus-libretro and the default gfx plugin (Glide64), Mario reacts on the 3rd frame, not the 2nd.

But more importantly, the angrylion video plugin acts the exact same way as GLupeN64, he jumps on the 5th frame. We've had a lengthy discussion on this before: https://github.com/GLupeN64/GLupeN64/issues/55

Basically, the fact that it works this way with angrylion (the most accurate gfx plugin we have) makes me comfortable with how it works here. This is also how it works in upstream mupen64plus (https://github.com/GLupeN64/GLupeN64/issues/55#issuecomment-249905286), so it's not an issue with RetroArch or this core. There may be something wrong with mupen64plus-core (https://github.com/mupen64plus/mupen64plus-core), and if that is the case, then someone will need to investigate that.

**Comment by Tasosgemah** [comment]
That extra frame you get, maybe your monitor is a bit slower? Or maybe you are using a very demanding shader? I heard that some shaders can also add some lag.

I re-tested on my CRT monitor. Did all the N64 cores. The results are:

Regular Mupen64 - Mario jumps on the second frame advance press.
ParaLLEl - Mario Jumps on the fifth frame advance press.
GlupeN64 - Mario Jumps on the fifth frame advance press.

So yeah. GlupeN64 is at the same level as ParaLLEl. Which is the same level as Mednafen Saturn. But that makes those three cores the slowest ones in RA library. And 3 or 4 frames slower than the real console (even on a CRT monitor). All the other cores give me 1, 2 or 3 frames. Personally, i wouldn't hold ParaLLEl as the standard since it's also in a very preliminary state. But that's just my opinion though because i would like Glupen to by my default core so maybe my expectations were unrealistic. And i don't know about programming so i don't know how possible it is to shave off a couple of frames in order to be a little more responsive.

Unfortunately, i don't know how to do a similar test with PJ64 + all these different plugins.

Thanks for the responses :)

**Comment by Tasosgemah** [comment]
Ok i did some further testing.

With the regular Muper64 core, If i change "buffer swap" to off i get ZERO lag, Mario jumps at the exact next frame.

I get the same results with GlupeN64 if i turn off Buffer emulation. That's ZERO lag right there. So there's must be something wrong with buffer emulation?

**Comment by loganmc10** [comment]
ParaLLEl, angrylion, and glupen64 all have 5 "frames" (at 60FPS) of delay, at least with Mario 64. I suspect CEN64 would be the same, although I don't think it has a frame advance feature.

It's not fair to say that something must be wrong with framebuffer emulation because there is a delay, perhaps there is something wrong without it because there is no delay.

I've never seen any sort of "scientific" evidence of there being no delay on a real N64 + CRT, so unless someone can demonstrate this on CEN64 (which is supposed to be more accurate), or a real N64, no one is going to look into it, since there might not even be a problem there.

**Comment by loganmc10** [comment]
Also, if it proves to be a real problem, it would not be an issue with this core, it would most likely be an issue with https://github.com/mupen64plus/mupen64plus-core.

**Comment by Tasosgemah** [comment]
Well, on the real console it definitely _feels_ more responsive than 5 frames. It feels immediate. And that's on a PAL console even (25fps).

I get the same results with the Saturn for instance. Playing Duke Nukem feels unresponsive with 5 frames lag (Mednafen Saturn) but on the real Saturn it feels great. 

Sure, i can't prove it without some actual test results but there is no way that the real console gives 5 frames of lag.

**Comment by loganmc10** [comment]
One other thing by the way, it's not 5 frames of lag. Mario jumps on the 3rd frame. RetroArch's frame advance feature moves at 60FPS even if the game is at 30FPS

You'll notice that in 2 of the frames nothing happens, they are duplicates.

so it goes like this:

you pause and hold down jump
1 Frame - things happen but Mario doesn't jump
2 Frame - Fake/duped frame
3 Frame - things happen but Mario doesn't jump
4 Frame - Fake/duped frame
5 Frame - Mario is in the air

So it would be more accurate to say there is a 3 frame (or 2 frame? I don't know how you count that) delay, since it is a 30FPS game.

**Comment by Tasosgemah** [comment]
"Also, if it proves to be a real problem, it would not be an issue with this core, it would most likely be an issue with https://github.com/mupen64plus/mupen64plus-core."

But the regular Mupen64plus doesn't give me lag. This lag is only present in ParaLLEl or Glupen with framebuffer ON. So it's not a problem with either Glupen or Mupen. Seems like a problem with Framebuffer emulation or GlideN64 to me.

**Comment by loganmc10** [comment]
If you use upstream mupen64plus with angrylion or GLideN64 you will experience the same thing

**Comment by Tasosgemah** [comment]
Ok i did a re-test. CRT monitor at 60hz, Vsync-ON. Hard GPU-ON. I also rotate the camera to see when a frame is repeated. "K" is the frame advance button.


Regular Mupen 64:

First press: Nothing happens, duplicated frame.
Second press: Camera rotates 1 frame. Mario is on the air.


GlupeN64 (Framebuffer ON)

First Press: Camera rotates 1 frame. Mario is on the ground.
Second press: Nothing happens, duplicated frame.
Third press Camera rotates 1 frame. Mario is on the ground.
Fourth press: Nothing happens, duplicated frame.
Fifth Press: Camera rotates 1 frame. Mario is on the air.

**Comment by loganmc10** [comment]
I would agree with the results of that test.

In conclusion:
It would be more accurate to say there is 2 frames of lag. If Mario was in the air on the first frame, a person would say there is "no lag", therefore we eliminate the frame that Mario is in the air. So there are 2 frames where Mario doesn't respond.

Secondly, angrylion is a "pixel perfect" software renderer, and has the exact same result, GLideN64 is the most accurate HLE OpenGL renderer, and has the same result. The 2 plugins were developed independently, using 2 different methods for rendering, and both came to the same conclusion. So I cannot assume the problem lies within the graphics plugin.

**Comment by Tasosgemah** [comment]
I agree that there are 2 frames of lag. But because it's a 30fps game it makes it worse because there 2 frames have to be repeated. But yeah, i agree.

I don't know where the problem lies. All i can say is that on the real console that i have next to me, Mario feels as responsive as he does on the regular mupen core. So i don't see a problem with the regular mupen there. The problem lies somewhere in Glupen and ParaLLEl. I don't know if it's the plugin or if Mupen behaves differently with those. All i know is that the regular Mupen core feels as responsive as the real thing.

**Comment by Tasosgemah** [comment]
Also, it's interesting that the first frames on Mupen or Glupen are different. One gets the repeated frame first, the other moves on the first frame.

**Comment by loganmc10** [comment]
I'm 90% sure it's related to this:

https://github.com/gonetz/GLideN64/issues/1124#issuecomment-247506760

https://github.com/gonetz/GLideN64/issues/1124#issuecomment-247509392

https://github.com/gonetz/GLideN64/issues/1124#issuecomment-248550602

I personally don't know enough about the inner working of the emulator core to know exactly where the problem probably lies, but basically it boils down to this:

GFX plugins that have framebuffer emulation (angrylion , paraLLEl, GLideN64) retrieve information (gonetz called it the "VI origin register", or "color buffer address register"), that is ignored when FB emulation is off/not used.

I don't doubt that there is input lag that shouldn't be there, but what I am saying that there is nothing wrong with the GFX plugins, they just render the scene given to them by the emulator. There is most likely something wrong with mupen64plus-core (the upstream project), and probably Project64, where they are giving an old/wrong scene to the GFX plugin to render when framebuffer emulation is happening.

**Comment by Tasosgemah** [comment]
Thanks for all the responses. I hope it gets sorted out some time, whatever the problem is.

**Comment by loganmc10** [comment]
By the way, this seems to be a Mario 64 specific problem. I tested both Super Smash Bros and Zelda: OoT and GLupeN64 and mupen64plus-libretro behave the same (in Smash Bros they "punch" on the 2nd frame and in Zelda Link's feet move on the.... 15th frame when I move forward, but it is the same in both emulators)

**Comment by Tasosgemah** [comment]
Hmmm, interesting.

I tested Yoshi's Story since its a 60fps platform game.

In Mupen core Yoshi jumps on the 3rd frame.
In Glupen Yoshi jumps on the 4th frame.


The difference is smaller than in Mario 64 but it's still slower for me. I wonder what Mario does so differently that gives such a result. Maybe it's the difference in effects/graphics that are emulated?

**Comment by Tasosgemah** [comment]
I also tested Smash Bros. And got the same results as you, both emulators react on second frame.

I'm afraid i don;t have enough knowledge to understand why there are all these differences between different games...

**Comment by loganmc10** [comment]
Yep I tested Yoshi's Story and got the same result as you lol.

gonetz (GLideN64 author) said this:

> Usually game uses two color buffers - one for rendering, other one displayed on screen. Game swaps buffers when rendering finished. Some games use triple buffering: one buffer for rendering, one waits to be displayed and one displayed on screen. Triple buffering usually use games running at 60 fps.

I assume that has something to do with it? I sounds like different games use different methods for buffer swapping, and I presume that is the difference we are seeing, but like you, I'm not knowledgeable in this area to really know for sure, hopefully someone from mupen64plus-core responds to the issue I filed there and they might be able to shed more light on the issue.
